src/contact.rs

```diff
@@ -0,0 +1,1120 @@
+                Ok(contact)
+            }
+        )
+    }
+
+    /// Returns `true` if this contact is blocked.
+    pub fn is_blocked(&self) -> bool {
+        self.blocked
+    }
+
+    /// Check if a contact is blocked.
+    pub fn is_blocked_load(context: &'a Context, id: u32) -> bool {
+        Self::load_from_db(context, &context.sql, id)
+            .map(|contact| contact.blocked)
+            .unwrap_or_default()
+    }
+
+    /// Block the given contact.
+    pub fn block(context: &Context, id: u32) {
+        set_block_contact(context, id, true);
+    }
+
+    /// Unblock the given contact.
+    pub fn unblock(context: &Context, id: u32) {
+        set_block_contact(context, id, false);
+    }
+
+    /// Add a single contact as a result of an _explicit_ user action.
+    ///
+    /// We assume, the contact name, if any, is entered by the user and is used "as is" therefore,
+    /// normalize() is _not_ called for the name. If the contact is blocked, it is unblocked.
+    ///
+    /// To add a number of contacts, see `dc_add_address_book()` which is much faster for adding
+    /// a bunch of addresses.
+    ///
+    /// May result in a `#DC_EVENT_CONTACTS_CHANGED` event.
+    pub fn create(context: &Context, name: impl AsRef<str>, addr: impl AsRef<str>) -> Result<u32> {
+        ensure!(
+            !addr.as_ref().is_empty(),
+            "Cannot create contact with empty address"
+        );
+
+        let (contact_id, sth_modified) =
+            Contact::add_or_lookup(context, name, addr, Origin::ManuallyCreated)?;
+        let blocked = Contact::is_blocked_load(context, contact_id);
+        context.call_cb(
+            Event::CONTACTS_CHANGED,
+            (if sth_modified == Modifier::Created {
+                contact_id
+            } else {
+                0
+            }) as uintptr_t,
+            0 as uintptr_t,
+        );
+        if blocked {
+            Contact::unblock(context, contact_id);
+        }
+
+        Ok(contact_id)
+    }
+
+    /// Mark all messages sent by the given contact
+    /// as _noticed_.  See also dc_marknoticed_chat() and dc_markseen_msgs()
+    ///
+    /// Calling this function usually results in the event `#DC_EVENT_MSGS_CHANGED`.
+    pub fn mark_noticed(context: &Context, id: u32) {
+        if sql::execute(
+            context,
+            &context.sql,
+            "UPDATE msgs SET state=? WHERE from_id=? AND state=?;",
+            params![DC_STATE_IN_NOTICED, id as i32, DC_STATE_IN_FRESH],
+        )
+        .is_ok()
+        {
+            context.call_cb(Event::MSGS_CHANGED, 0, 0);
+        }
+    }
+
+    /// Check if an e-mail address belongs to a known and unblocked contact.
+    /// Known and unblocked contacts will be returned by `dc_get_contacts()`.
+    ///
+    /// To validate an e-mail address independently of the contact database
+    /// use `dc_may_be_valid_addr()`.
+    pub fn lookup_id_by_addr(context: &Context, addr: impl AsRef<str>) -> u32 {
```

Comment:
This function seems to return 0 as failure.  I think this would be better if it returned Result.